Fix bled112 hangs when dongle removed and allow for checking dongle functionality from HardwareManager

timburke opened the pull request:
This PR fixes some critical bugs inside the BLED112 device adapter and makes it easier to build applications that need to know when there is a bled112 dongle removed during the application running.

The major changes are:
1. Refactors how serial port exceptions are handled inside the background command loop to ensure that `HardwareManager` commands on a bled112 dongle never hang.  Previously if there was a serial communication error, it would not trigger a failure callback.
2. Adds a new HardwareManager api, `heartbeat`.  This is implemented currently in the `bled112`, `virtual` and `emulated` device adapters and checks if the underlying bled112 dongle is functioning.  It returns True if things are good and False is there's a problem.  The purpose is to allow for polling to see if a dongle has died.
3. Adds a background event propagation mechanism so that the `heartbeat` command will properly read `False` even if you are just `scanning` with the bled112 dongle.  Now serial exceptions in the background reader thread are translated into a sentinal bgapi event `255, 255` that means hardware communication failure.  This triggers the `periodic_callback` to mark the DeviceAdapter as dead within 1 second (on the next `periodic_callback` invocation).

**Testing Performed**
1. I verified that connect and connect_direct no longer hang when the dongle is removed.
2. I verified no exceptions or hangs on exit from `iotile` if a dongle is removed
3. I verified that `iotile hw heartbeat` works and detects dongle removal both when connected to devices and when just scanning.

All tests were performed on Windows 10.

@JKHeadley Can you checkout this branch and see if it provides the APIs you need for your use case?

Closes #988 
Closes #989

x24ling commented:
I ran the test Tim describes and they all look like they pass for me. I have also verified that action 2. causes the bled112 to hang in the previous version in bled112, and it works.

Knowing that this hasn't been implemented in the other transport_link package, jlink, this comment may be low priority, but these are the errors I get when I call heartbeat in jlink aren't very descriptive.

```
iotile hw --port jlink:mux=ftdi
(HardwareManager) heartbeat
DeviceAdapterError: Operation none on conn 0 failed because connection in wrong state
Additional Information:
reason: connection in wrong state
operation: none
conn_id: 0
(HardwareManager) connect_direct 'channel=0;device=nrf52'
(HardwareManager) controller tile_version back
[4, 1, 2]
(HardwareManager) heartbeat
Traceback (most recent call last):
  File "c:\users\chris\iotile\coretools\iotilecore\iotile\core\scripts\iotile_script.py", line 244, in main
    finished = shell.invoke_string(linebuf)
  File "c:\users\chris\iotile\env\new_prod_workspace\lib\site-packages\typedargs\shell.py", line 518, in invoke_string
    return self.invoke(args)
  File "c:\users\chris\iotile\env\new_prod_workspace\lib\site-packages\typedargs\shell.py", line 489, in invoke
    val, line, finished = self.invoke_one(line)
  File "c:\users\chris\iotile\env\new_prod_workspace\lib\site-packages\typedargs\shell.py", line 452, in invoke_one
    val = func(*posargs, **kwargs)
  File "c:\users\chris\iotile\coretools\iotilecore\iotile\core\hw\hwmanager.py", line 315, in heartbeat
    result = self.stream.debug_command('heartbeat')
  File "c:\users\chris\iotile\coretools\iotilecore\iotile\core\hw\transport\adapterstream.py", line 465, in debug_command
    return self._loop.run_coroutine(self.adapter.debug(0, cmd, args))
  File "c:\users\chris\iotile\coretools\iotilecore\iotile\core\utilities\async_tools\event_loop.py", line 540, in run_coroutine
    return future.result()
  File "C:\Python37\lib\concurrent\futures\_base.py", line 432, in result
    return self.__get_result()
  File "C:\Python37\lib\concurrent\futures\_base.py", line 384, in __get_result
    raise self._exception
  File "c:\users\chris\iotile\env\new_prod_workspace\lib\site-packages\iotile_transport_jlink\jlink.py", line 288, in debug
    return await func(self._device_info, self._control_info, cmd_args)
TypeError: 'NoneType' object is not callable
(HardwareManager)      
```

timburke commented:
Thanks for the testing @x24ling.  I just updated the PR with placeholder support for `heartbeat` in the jlink adapter as well.  Now it should report `true` just like the emulated adapter.

JKHeadley commented:
@timburke I'm not sure if I've installed the PR correctly but I am still getting hangs when I follow the steps in #988 . My pip list seems to show that I'm using the correct versions:
```
iotile-core                     5.1.2     /Users/jheadley/Documents/GitHub/coretools/iotilecore              
iotile-transport-bled112        3.0.6     /Users/jheadley/Documents/GitHub/coretools/transport_plugins/bled112
```

Is it possible that my iotile commands are using a previous version?

JKHeadley commented:
@timburke heartbeat does not work with an emulated port:
```
iotile hw --port=emulated:reference_1_0
(HardwareManager) heartbeat
DeviceAdapterError: Operation none on conn 0 failed because connection id does not exist
Additional Information:
reason: connection id does not exist
operation: none
conn_id: 0
```

timburke commented:
@JKHeadley can you send logs with -vvvv when you reproduce the hang using the iotile tool.

That will let me either debug or see that the code wasn't installed.

For the emulated device you need to install the updated iotile-emulate

JKHeadley commented:
@timburke here is the output:

```
iotile -vvvv hw --port=bled112
20-07-16 09:42:18.925 DEB iotile_transport_bled112.bled112 Found BLED112 device at /dev/cu.usbmodem11
20-07-16 09:42:18.931 INF iotile_transport_bled112.bled112 Using first available BLED112 adapter at /dev/cu.usbmodem11
20-07-16 09:42:18.931 WAR iotile_transport_bled112.bled112 pycryptodome is not installed, encrypted v2 broadcasts will be dropped.
20-07-16 09:42:19.436 INF iotile_transport_bled112.bled112 BLED112 adapter supports 3 connections
20-07-16 09:42:19.441 DEB iotile.core.utilities.async_tools.event_loop Starting event loop
20-07-16 09:42:19.442 DEB asyncio Using selector: KqueueSelector
20-07-16 09:42:19.442 DEB iotile.core.utilities.async_tools.event_loop Starting loop in background thread
20-07-16 09:42:19.442 DEB iotile.core.utilities.async_tools.event_loop Added primary task LegacyAdapter (BLED112Adapter)
20-07-16 09:42:19.443 DEB iotile.core.utilities.async_tools.event_loop Added subtask periodic task for BLED112Adapter to parent LegacyAdapter (BLED112Adapter)
(HardwareManager) 20-07-16 09:42:38.058 ERR iotile_transport_bled112.async_packet Error in reader thread, putting device failure event
Traceback (most recent call last):
  File "/Users/jheadley/Documents/GitHub/iotile-factory-companion/electron/factory_companion_server/venv/lib/python3.7/site-packages/serial/serialposix.py", line 493, in read
    buf = os.read(self.fd, size - len(read))
OSError: [Errno 6] Device not configured

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jheadley/Documents/GitHub/coretools/transport_plugins/bled112/iotile_transport_bled112/async_packet.py", line 68, in reader_thread
    chunk = bytearray(filelike.read(header_length - len(header)))
  File "/Users/jheadley/Documents/GitHub/iotile-factory-companion/electron/factory_companion_server/venv/lib/python3.7/site-packages/serial/serialposix.py", line 509, in read
    raise SerialException('read failed: {}'.format(e))
serial.serialutil.SerialException: read failed: [Errno 6] Device not configured
20-07-16 09:42:38.087 WAR iotile_transport_bled112.bled112 Detected that the dongle was broken or disconnected
20-07-16 09:42:38.529 WAR iotile_transport_bled112.bled112 Hardware failure detected, cleanly stopping device adapter

(HardwareManager) connect 0x053d
20-07-16 09:43:02.323 INF iotile.core.hw.transport.adapter.async_wrapper Inside connect, conn_id=0, conn_string=E4:9C:4E:0D:CC:5A
```

The prompt never returns after the last line

JKHeadley commented:
@timburke for `iotile-emulate` here is what I've tried:

Install the package from this PR:
```
pip install -e ~/Documents/GitHub/coretools/iotileemulate
```

Try to use `heartbeat` with an emulated port:
```
iotile -vvvv hw --port=emulated:reference_1_0      
20-07-16 09:51:18.403 DEB iotile.core.utilities.async_tools.event_loop Starting event loop
20-07-16 09:51:18.404 DEB asyncio Using selector: KqueueSelector
20-07-16 09:51:18.405 DEB iotile.core.utilities.async_tools.event_loop Starting loop in background thread
20-07-16 09:51:18.409 DEB iotile.emulate.internal.rpc_queue Starting RPC dispatch task
20-07-16 09:51:18.409 INF iotile.emulate.virtual.emulated_tile Resetting controller
20-07-16 09:51:18.409 INF iotile.emulate.virtual.emulated_tile Tile at address 8 has reset itself.
20-07-16 09:51:18.409 INF iotile.emulate.virtual.emulated_tile Starting main task for tile at address 8
20-07-16 09:51:18.410 INF iotile.emulate.virtual.emulated_tile Latched config variables at reset for controller: {'fast_tick': 0, 'user_tick_1': 0, 'user_tick_2': 0, 'storage_fillstop': False, 'streaming_fillstop': False}
20-07-16 09:51:18.411 INF iotile.emulate.reference.controller_features.sensor_log Config vars in sensor log reset: {'fast_tick': 0, 'user_tick_1': 0, 'user_tick_2': 0, 'storage_fillstop': False, 'streaming_fillstop': False}
20-07-16 09:51:18.412 DEB iotile.emulate.reference.controller_features.sensor_graph sensor_graph subsystem task starting
20-07-16 09:51:18.412 INF iotile.emulate.virtual.emulated_tile Finished clearing controller to reset condition
(HardwareManager) 
(HardwareManager) 
(HardwareManager) heartbeat
DeviceAdapterError: Operation none on conn 0 failed because connection id does not exist
Additional Information:
reason: connection id does not exist
operation: none
conn_id: 0
(HardwareManager) 
```

timburke commented:
@JKHeadley Thank you for checking and reporting these two issues.

I understand the `iotile-emulate` bug now and will fix it presently.

I'll dig into the hang bug.  Based on your logs I should be able to solve it quickly as well and then update the PR for you to recheck and make sure it's smashed.

timburke commented:
@JKHeadley I believe I have fixed the hang on connect issue, could you please recheck with the latest code.

## Tests Performed

**Dongle Fails and Then Connect**

Test case: unplug dongle, run `connect UUID` to a device that was previously scanned.
Result: No hang

```
(env) PS C:\Arch\core\coretools> iotile -vvvv hw --port=bled112
20-07-16 10:34:45.641 DEB iotile_transport_bled112.bled112 Found BLED112 device at COM4
20-07-16 10:34:45.642 DEB iotile_transport_bled112.bled112 Found BLED112 device at COM5
20-07-16 10:34:45.646 DEB iotile_transport_bled112.bled112 Can't use BLED112 device COM4 because it's locked
20-07-16 10:34:45.652 INF iotile_transport_bled112.bled112 Using first available BLED112 adapter at COM5
20-07-16 10:34:46.165 INF iotile_transport_bled112.bled112 BLED112 adapter supports 3 connections
20-07-16 10:34:46.171 DEB iotile.core.utilities.async_tools.event_loop Starting event loop
20-07-16 10:34:46.172 DEB asyncio Using proactor: IocpProactor
20-07-16 10:34:46.176 DEB iotile.core.utilities.async_tools.event_loop Starting loop in background thread
20-07-16 10:34:46.176 DEB iotile.core.utilities.async_tools.event_loop Added primary task LegacyAdapter (BLED112Adapter)
20-07-16 10:34:46.179 DEB iotile.core.utilities.async_tools.event_loop Added subtask periodic task for BLED112Adapter to parent LegacyAdapter (BLED112Adapter)

(HardwareManager) 20-07-16 10:34:59.397 ERR iotile_transport_bled112.async_packet Error in reader thread, putting device failure event
Traceback (most recent call last):
  File "c:\arch\core\coretools\transport_plugins\bled112\iotile_transport_bled112\async_packet.py", line 68, in reader_thread
    chunk = bytearray(filelike.read(header_length - len(header)))
  File "c:\arch\core\workspace\env\lib\site-packages\serial\serialwin32.py", line 273, in read
    raise SerialException("ClearCommError failed ({!r})".format(ctypes.WinError()))
serial.serialutil.SerialException: ClearCommError failed (PermissionError(13, 'The device does not recognize the command.', None, 22))
20-07-16 10:34:59.510 WAR iotile_transport_bled112.bled112 Detected that the dongle was broken or disconnected

(HardwareManager) connect 1
20-07-16 10:35:07.395 INF iotile.core.hw.transport.adapter.async_wrapper Inside connect, conn_id=0, conn_string=88:6B:0F:66:A7:38
20-07-16 10:35:07.395 WAR iotile_transport_bled112.bled112_cmd Command ['_stop_scan'] sent after background processor was stopped; failing immediately
HardwareError: Synchronous command ['_stop_scan'] failed because background processor was stopped
(HardwareManager)
```

**Dongle Fails During Connect**

Test Case: Unplug POD so that connection attempts will take 4s to timeout.  Run `connect UUID`.  Unplug dongle.
Result: No hang, 4 second timeout and then a timeout error raised.

```
(HardwareManager) connect 1                                                                                                                           
20-07-16 10:36:09.155 INF iotile.core.hw.transport.adapter.async_wrapper Inside connect, conn_id=0, conn_string=88:6B:0F:66:A7:38                     
20-07-16 10:36:10.476 ERR iotile_transport_bled112.async_packet Error in reader thread, putting device failure event                                  
Traceback (most recent call last):                                                                                                                      
File "c:\arch\core\coretools\transport_plugins\bled112\iotile_transport_bled112\async_packet.py", line 68, in reader_thread                             
	chunk = bytearray(filelike.read(header_length - len(header)))                                                                                       
File "c:\arch\core\workspace\env\lib\site-packages\serial\serialwin32.py", line 273, in read                                                            
	raise SerialException("ClearCommError failed ({!r})".format(ctypes.WinError()))                                                                   
	serial.serialutil.SerialException: ClearCommError failed (PermissionError(13, 'The device does not recognize the command.', None, 22))                
20-07-16 10:36:10.479 WAR iotile_transport_bled112.bled112 Detected that the dongle was broken or disconnected                                        
20-07-16 10:36:10.770 WAR iotile_transport_bled112.bled112 Hardware failure detected, cleanly stopping device adapter                                 
DeviceAdapterError: Operation connect on conn 0 failed because Timeout opening connection                                                             
Additional Information:                                                                                                                               
reason: Timeout opening connection                                                                                                                    
operation: connect                                                                                                                                    
conn_id: 0                                                                                                                                            
(HardwareManager)                                                                  
```

The underlying issue was due to how background scan errors were detected and fixed.  I will have a fix for the `iotile-emulate` heartbeat functionality soon as well.

timburke commented:
@JKHeadley Running `heartbeat` on the emulated adapter should now work as well with the latest commit:

```
(env) PS C:\Arch\core\coretools> iotile hw --port=emulated:reference_1_0
(HardwareManager) scan
{
    "connection_string": "1",
    "expiration_time": 723456.812,
    "signal_strength": 100,
    "slug": "d--0000-0000-0000-0001",
    "uuid": 1,
    "validity_period": 600000
}
(HardwareManager) heartbeat
True
```

JKHeadley commented:
@timburke this looks to have resolved all the issues I was experiencing, thanks!
